### PR TITLE
Min connection + jslint

### PR DESCRIPTION
**Pull request by guytuboul**

### CODE REVIEW

**Comment by argon** [comment]
Please fix the whitespace changes. While the formatting of the module is a little inconsistent I'm not going to merge any whitespace fixes as part of a feature pull request.

File: lib/connection.js [hunk 1]
```diff
@@ -69,39 +72,55 @@ function Connection (options) {
...
-	events.EventEmitter.call(this);
-}
+           
+        events.EventEmitter.call(this);
+        this.maintainMinConnection();
+}   
 
 sysu.inherits(Connection, events.EventEmitter);
+Connection.prototype.maintainMinConnection = function() {
+     var that = this;
```
Comment (argon) [inline lib/connection.js]: I don't approve of `var that = this` or similar. Please bind `createConnections` to `this`

**Comment by argon** [comment]
There is a lot of unnecessary complexity here, I believe.

Why did you choose to create the connections in parallel? There appears to be quite a big problem at present: You are creating a new socket every 100ms, but you are testing against the length of `this.sockets` which will only be populated when the connection has *finished* establishing. TLS handshaking usually takes ~800ms based on my testing. If I configured the module with `minConnections = 2` and `maxConnections = 5`, by the time the **first** connection is established the system will have ~8 connections being created.

It would be far simpler to keep the sequential connection creation, then call `maintainMinConnections` after each connection has established which should simply check whether `this.sockets.length < this.options.minConnections` and call `createConnection` again if necessary.

**Comment by guytuboul** [comment]
1)I will fix the whitespace changes, and bind createConnections to this
2)I don't agree with you :). I don't create the connections in parallel. I am only creating connection if:
"if (that.deferredConnection === null)", meaning there is not other connection in progress.
Every 100ms (We can change this interval) i am checking if i have green light to create another connection.

**Comment by argon** [comment]
I see, in that case why have the timeout at all? You could simply check whether another connection needs to be created after each connection resolves.

**Comment by guytuboul** [comment]
Yes, i agree with that. i will change that (probably tomorrow).
Do you know why the test are failing?

**Comment by argon** [comment]
You have changed the behaviour such that it immediately attempts to establish a connection when the instance is created. The tests assume that the connection won't be established until the first notification is sent.

**Comment by guytuboul** [comment]
I changed my pull request, you can review.
About the test that fails now,can you take care of this?

File: lib/connection.js [hunk 2]
```diff
@@ -196,6 +219,7 @@ Connection.prototype.createSocket = function () {
 		this.socket.on("data",    this.handleTransmissionError.bind(this, this.socket));
 		this.socket.on("drain",   this.socketDrained.bind(this, this.socket, true));
 		this.socket.once("close", this.socketClosed.bind(this, this.socket));
+                this.socket.deferredConnection = this.deferredConnection;
```
Comment (argon) [inline lib/connection.js]: Why do you retain a copy of the deferred connection on the socket object here?

File: lib/connection.js [hunk 3]
```diff
@@ -225,11 +249,15 @@ Connection.prototype.createSocket = function () {
 /**
  * @private
  */
+Connection.prototype.createNewConnection = function() {
+    if (this.deferredConnection === null && this.sockets.length < this.options.maxConnections) {
```
Comment (argon) [inline lib/connection.js]: I think this method adds unnecessary complexity. The wrapper is not needed. The only other place `createConnection` is called directly (bypassing this check) is on 124 were you first check whether the max number of connections has been reached before then checking whether the number of sockets is less than the minimum. By changing `maintainMinConnection` to

```
Connection.prototype.maintainMinConnection = function() {
    if (typeof (this.options.minConnections - this.sockets.length) === 'number') {
        if (this.options.minConnections > this.sockets.length) {
            return;
        }
        this.createNewConnection();
    }
}
```

File: lib/connection.js [hunk 4]
```diff
@@ -99,10 +104,28 @@ function Connection (options) {
...
 
 sysu.inherits(Connection, events.EventEmitter);
 
+
+/**
+ *
+ * @private
+ */
+Connection.prototype.maintainMinConnection = function() {
+    if (typeof (this.options.minConnections - this.sockets.length) === 'number') {
```
Comment (argon) [inline lib/connection.js]: What is this check guarding against? We know that this.options.minConnections will have a numeric value because it's setup in the initialiser.

File: lib/connection.js (hunk 2 above)
Comment (guytuboul) [inline lib/connection.js]: i am using this in Connection.prototype.errorOccurred : 
if (socket.deferredConnection && socket.deferredConnection.promise.isPending()) {
           socket.deferredConnection.reject(err);
}

File: lib/connection.js (hunk 4 above)
Comment (guytuboul) [inline lib/connection.js]: agree,it was necessary in previous code

File: lib/connection.js (hunk 2 above)
Comment (argon) [inline lib/connection.js]: Yes, I see that, but it was `this.deferredConnection` before, why did you change it? I don't understand how it's relevant to the min connections feature

Comment (guytuboul) [inline lib/connection.js]: You right again:) i miss something in your code. 
This is my last try:)

**Comment by guytuboul** [comment]
Hey, so what is the status of this pull request? You approved it?(we still need to change the tests..)
Do you have time to fix the test?

**Comment by argon** [comment]
I haven't approved it, there are still several problems now that I am able to run the code. If you're happy for me to do so I will fix them and pull your code in. If you would like to fix them then I can provide more details. Basically the round-robin code does not do the right thing. Also when `shutdown` is used the minConnections code doesn't stop creating sockets so it causes the process to hang instead of closing everything down and exiting.

**Comment by guytuboul** [comment]
Yes,if you want to change the code,its ok :).
I will try to fix this issues tomorrow morning

**Comment by guytuboul** [comment]
What is the problem with the round robin?

**Comment by argon** [comment]
The for loop it is in decides which socket to check. If you're going to pick a socket based on other criteria then the for loop should be changed. While I don't disagree with the idea of using a round robin I think it's potentially quite complicated. Can't see an elegant solution around it.

―A

**Comment by guytuboul** [comment]
Why not using round robin? This is why you want to have min connection, so
you won't try to send the notification on one socket....

**Comment by argon** [comment]
I understand the purpose, I'm just not sure it's the best solution. Still thinking about this.

—A

**Comment by argon** [comment]
Sending notifications on a single socket isn't the problem here, that's taken care of by the socket availability. The only reason two sockets would be faster than one is if one is blocked which the current checking takes care of. The value in having a minimum number of connections is that if one connection drops you don't need to wait the ~800ms to recreate the socket, there's already one there.

The value in a round robin is to make sure both connections stay alive. Having thought about it, the following would be a good replacement.

var socket = this.sockets[(i + this.connectionRoundRobin) this.sockets.length]

Or something like that.

—A

**Comment by guytuboul** [comment]
I know that apple recommend to send the notification on multiple
connections(i can send you the doc).
When apple recommend something,there is a chance that you MUST use it :).
I am working in glide,and we send dozens millions of notification per
dat,and we recently have a drop on push that reach the destination,and we
think that this can help us..
I will change the things that we discuss about tomorrow...

**Comment by guytuboul** [comment]
Why do you think that var socket = this.sockets[(i +
this.connectionRoundRobin) this.sockets.length] is better?

**Comment by guytuboul** [comment]
var socket = this.sockets[(i +
this.connectionRoundRobin)%this.sockets.length]

**Comment by guytuboul** [comment]
I pushed the fix for the not creating connection if shutdownPending is true.
Can you take care of the tests that fails

File: lib/connection.js [hunk 5]
```diff
@@ -80,7 +82,8 @@ function Connection (options) {
 	// Set cache length to 1 to ensure transmitted notifications can be sent.
 	this.options.cacheLength = Math.max(this.options.cacheLength, 1);
 	this.options.maxConnections = Math.max(this.options.maxConnections, 1);
-
+        this.options.minConnections = Math.max(this.options.minConnections, 1);
```
Comment (argon) [inline lib/connection.js]: Please use tabs not spaces.

File: lib/connection.js [hunk 6]
```diff
@@ -80,7 +82,8 @@ function Connection (options) {
...
+
```
Comment (argon) [inline lib/connection.js]: No leading whitespace please

File: lib/connection.js [hunk 7]
```diff
@@ -88,7 +91,9 @@ function Connection (options) {
 	this.socketId = 0;
 
 	this.failureCount = 0;
-
+        
+        this.currentConnectionRoundRobin = 0;
+
```
Comment (argon) [inline lib/connection.js]: Tabs not spaces and remove leading whitespace

File: lib/connection.js [hunk 8]
```diff
@@ -99,10 +104,22 @@ function Connection (options) {
 	this.terminated = false;
 
 	events.EventEmitter.call(this);
+        
+        this.maintainMinConnection();
```
Comment (argon) [inline lib/connection.js]: Tabs not spaces, leading whitespace.

File: lib/connection.js [hunk 9]
```diff
@@ -99,10 +104,22 @@ function Connection (options) {
...
+
+/**
+ *
+ * @private
+ */
+Connection.prototype.maintainMinConnection = function() {
+    if (this.sockets.length < this.options.minConnections && !this.shutdownPending) {
+        this.createConnection();
+    }
+}
```
Comment (argon) [inline lib/connection.js]: Tabs not spaces.

File: lib/connection.js [hunk 10]
```diff
@@ -226,7 +243,7 @@ Connection.prototype.createSocket = function () {
  * @private
  */
 Connection.prototype.createConnection = function() {
-	if (this.initialisingConnection() || this.sockets.length >= this.options.maxConnections) {
+    	if (this.deferredConnection !== null || this.sockets.length >= this.options.maxConnections) {
```
Comment (argon) [inline lib/connection.js]: I think this line needs to be reverted now?

File: lib/connection.js [hunk 11]
```diff
@@ -269,6 +286,7 @@ Connection.prototype.createConnection = function() {
 		trace("create completed", this.sockets.length);
 		this.deferredConnection = null;
 		this.socket = undefined;
+                this.maintainMinConnection();
```
Comment (argon) [inline lib/connection.js]: Tabs.

File: lib/connection.js [hunk 12]
```diff
@@ -299,7 +317,8 @@ Connection.prototype.initialisingConnection = function() {
 	do {
 		socketsAvailable = 0;
 		for (var i = 0; i < this.sockets.length; i++) {
-			socket = this.sockets[i];
+                        socket = this.sockets[this.currentConnectionRoundRobin % this.sockets.length];
+                        this.currentConnectionRoundRobin += 1;
```
Comment (argon) [inline lib/connection.js]: Tabs.

File: lib/connection.js [hunk 13]
```diff
@@ -457,6 +476,7 @@ Connection.prototype.socketClosed = function(socket) {
 	if (index > -1) {
 		this.sockets.splice(index, 1);
 	}
+        this.maintainMinConnection();
```
Comment (argon) [inline lib/connection.js]: Tabs

File: lib/connection.js [hunk 14]
```diff
@@ -745,4 +765,4 @@ Connection.prototype.shutdown = function () {
 	this.shutdownPending = true;
 };
 
-module.exports = Connection;
+module.exports = Connection;
```
Comment (argon) [inline lib/connection.js]: Please leave the newline at the end of the file.

File: lib/connection.js [hunk 15]
```diff
@@ -53,6 +54,7 @@ function Connection (options) {
 		cacheLength: 1000,
 		autoAdjustCache: true,
 		maxConnections: 1,
+                minConnections: 1,
```
Comment (argon) [inline lib/connection.js]: Tabs not spaces, please.

**Comment by argon** [comment]
I have attached my comments but I am happy to make those changes when I pull the code in.

File: lib/connection.js (hunk 10 above)
Comment (guytuboul) [inline lib/connection.js]: O.k,although i believe this is nicer:)

**Comment by guytuboul** [comment]
O.k all test passed:).Do you want to merge this?

**Comment by argon** [comment]
Sure, I'll merge this when I get home.

**Comment by guytuboul** [comment]
Great,thanks

**Comment by guytuboul** [comment]
Hi, i think that there is one optimization that we can do
Instead of  :
socket = this.sockets[this.currentConnectionRoundRobin %
this.sockets.length];
this.currentConnectionRoundRobin += 1;
we need to do :
                       socket =
this.sockets[this.currentConnectionRoundRobin];
this.currentConnectionRoundRobin = (this.currentConnectionRoundRobin + 1) %
this.sockets.length;

So this.currentConnectionRoundRobin  will not reach the sky with +=1... :)
Can you change to code?
Thanks
